### PR TITLE
SPSH-842

### PR DESCRIPTION
**Pull request by DeFi3298**
Backend Part for developing a TokenReset View for the TwoFactorAuthentification

[(https://ticketsystem.dbildungscloud.de/browse/SPSH-842)](https://ticketsystem.dbildungscloud.de/browse/SPSH-842)

### CODE REVIEW

**Review by marode-cap — CHANGES_REQUESTED** [review]

File: src/modules/privacy-idea-administration/privacy-idea-administration.service.ts [hunk 1]
```diff
@@ -198,4 +200,43 @@ export class PrivacyIdeaAdministrationService {
...
+        if (!twoAuthState) {
+            throw new Error('Error getting two-factor auth state.');
+        }
+        const serial: string = twoAuthState.serial;
+        try {
+            const response: ResetTokenResponse = await this.unassignToken(serial, token);
+            return response;
+        } catch (error) {
+            throw new Error(`Error initializing token: `);
+        }
```
[the remark this answers is not in the document]
Comment (DeFi3298) [reply src/modules/privacy-idea-administration/privacy-idea-administration.service.ts]: done

File: src/modules/privacy-idea-administration/privacy-idea-administration.service.spec.ts [hunk 2]
```diff
@@ -235,4 +235,91 @@ describe(`PrivacyIdeaAdministrationService`, () => {
...
     });
+    describe('resetToken', () => {
+        it('should reset token successfully', async () => {
+            const mockResetUser: string = 'testUser';
+            const mockJWTToken: string = 'mockJWTToken';
+            const mockTwoAuthState: ResetTokenPayload = createMock<ResetTokenPayload>();
+            const mockResetTokenResponse: ResetTokenResponse = createMock<ResetTokenResponse>();
+
+            jest.spyOn(service as unknown as { getJWTToken: () => Promise<string> }, 'getJWTToken').mockResolvedValue(
+                mockJWTToken,
```
[the remark this answers is not in the document]
Comment (DeFi3298) [reply src/modules/privacy-idea-administration/privacy-idea-administration.service.spec.ts]: done

File: src/modules/privacy-idea-administration/privacy-idea-administration.controller.ts [hunk 3]
```diff
@@ -89,4 +89,17 @@ export class PrivacyIdeaAdministrationController {
...
+    @HttpCode(HttpStatus.OK)
+    @ApiCreatedResponse({ description: 'The token was successfully reset.', type: PrivacyIdeaAdministrationService })
+    @ApiBadRequestResponse({ description: 'A username was not given or not found.' })
+    @ApiUnauthorizedResponse({ description: 'Not authorized to reset token.' })
+    @ApiForbiddenResponse({ description: 'Insufficient permissions to reset token.' })
+    @ApiNotFoundResponse({ description: 'Insufficient permissions to reset token.' })
+    @ApiInternalServerErrorResponse({ description: 'Internal server error while reseting a token.' })
+    @Public()
+    public async resetToken(@Query('userName') userName: string): Promise<ResetTokenResponse | undefined> {
+        return this.privacyIdeaAdministrationService.resetToken(userName);
```
[the remark this answers is not in the document]
Comment (DeFi3298) [reply src/modules/privacy-idea-administration/privacy-idea-administration.controller.ts]: done

File: src/modules/privacy-idea-administration/privacy-idea-administration.controller.spec.ts [hunk 4]
```diff
@@ -193,4 +193,64 @@ describe('PrivacyIdeaAdministrationController', () => {
             );
         });
     });
+    describe('PrivacyIdeaAdministrationController resetToken', () => {
+        it('should successfully reset a token', async () => {
+            const userName: string = 'user1';
+            const mockResetTokenResponse: ResetTokenResponse = createMock<ResetTokenResponse>();
+            jest.spyOn(serviceMock, 'resetToken').mockResolvedValue(mockResetTokenResponse);
+
```
[the remark this answers is not in the document]
Comment (DeFi3298) [reply src/modules/privacy-idea-administration/privacy-idea-administration.controller.spec.ts]: done

File: src/modules/privacy-idea-administration/privacy-idea-administration.service.ts [hunk 5]
```diff
@@ -198,4 +209,126 @@ export class PrivacyIdeaAdministrationService {
...
+        try {
+            const response: AxiosResponse<ResetTokenResponse> = await firstValueFrom(
+                this.httpService.post(url, payload, { headers }),
+            );
+            return response.data;
+        } catch (error) {
+            if (error instanceof Error) {
+                throw new Error(`Error unassigning token: ${error.message}`);
+            } else {
+                throw new Error(`Error unassigning token: Unknown error occurred`);
```
Comment (marode-cap) [inline src/modules/privacy-idea-administration/privacy-idea-administration.service.ts:330]: Should these be generic Errors?

**Review by marode-cap — APPROVED** [review]
Approved, but you should still check the three comments

**Pull request merged**